src/ripple/app/tx/impl/DID.cpp

```diff
@@ -0,0 +1,215 @@
+DIDSet::doApply()
+{
+    // Edit ledger object if it already exists
+    Keylet const didKeylet = keylet::did(account_);
+    if (auto const sleDID = ctx_.view().peek(didKeylet))
+    {
+        if (auto const uri = ctx_.tx[~sfURI])
+        {
+            if (uri->empty())
+            {
+                sleDID->makeFieldAbsent(sfURI);
+            }
+            else
+            {
+                (*sleDID)[sfURI] = *uri;
+            }
+        }
+        if (auto const data = ctx_.tx[~sfData])
+        {
+            if (data->empty())
+            {
+                sleDID->makeFieldAbsent(sfData);
+            }
+            else
+            {
+                (*sleDID)[sfData] = *data;
+            }
+        }
+        if (!sleDID->isFieldPresent(sfURI) && !sleDID->isFieldPresent(sfData))
+        {
+            return tecEMPTY_DID;
+        }
+        ctx_.view().update(sleDID);
+        return tesSUCCESS;
+    }
+
+    // Create new ledger object otherwise
+    auto const sleDID = std::make_shared<SLE>(didKeylet);
+    (*sleDID)[sfAccount] = account_;
+    if (auto const uri = ctx_.tx[~sfURI]; uri.has_value() && !uri->empty())
+        (*sleDID)[sfURI] = uri.value();
+    if (auto const data = ctx_.tx[~sfData]; data.has_value() && !data->empty())
+        (*sleDID)[sfData] = data.value();
+
+    if (auto const ret = addSLE(ctx_, sleDID, account_); !isTesSuccess(ret))
+        return ret;
+
+    return tesSUCCESS;
+}
+
+NotTEC
+DIDDelete::preflight(PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled(featureDID))
+        return temDISABLED;
+
+    if (ctx.tx.getFlags() & tfUniversalMask)
+        return temINVALID_FLAG;
+
+    if (auto const ret = preflight1(ctx); !isTesSuccess(ret))
+        return ret;
+
+    return preflight2(ctx);
+}
+
+TER
+DIDDelete::deleteSLE(ApplyContext& ctx, Keylet sleKeylet, AccountID const owner)
+{
+    auto const sle = ctx.view().peek(sleKeylet);
+    if (!sle)
+        return tecNO_ENTRY;
+
+    return DIDDelete::deleteSLE(ctx.view(), sle, owner, ctx.journal);
+}
+
+TER
+DIDDelete::deleteSLE(
+    ApplyView& view,
+    std::shared_ptr<SLE> sle,
+    AccountID const owner,
+    beast::Journal j)
+{
+    // Remove object from owner directory
+    {
+        auto const page = (*sle)[sfOwnerNode];
+        if (!view.dirRemove(keylet::ownerDir(owner), page, sle->key(), true))
+        {
+            JLOG(j.fatal()) << "Unable to delete DID Token from owner.";
+            return tefBAD_LEDGER;
+        }
+    }
+
+    auto const sleOwner = view.peek(keylet::account(owner));
+    adjustOwnerCount(view, sleOwner, -1, j);
+    view.update(sleOwner);
+
+    // Remove object from ledger
+    view.erase(sle);
+    return tesSUCCESS;
+}
+
+TER
+DIDDelete::doApply()
+{
+    AccountID const account = ctx_.tx[sfAccount];
+    auto const didKeylet = keylet::did(account);
+
+    if (auto const ret = deleteSLE(ctx_, didKeylet, account_);
+        !isTesSuccess(ret))
+        return ret;
+
+    return tesSUCCESS;
```

Comment:
Same comment as before: why a separate function instead of inlining the code?

Reply:
Ditto here